File: energielenker/energielenker/item_price/item_price.py
# -*- coding: utf-8 -*-
# Copyright (c) 2021, libracore and contributors
# For license information, please see license.txt

import frappe
from erpnext.manufacturing.doctype.bom.bom import BOM

@frappe.whitelist()
def update_bom_item_cost(name, item_code, rate, rm_cost_as_per, price_list=None):
    # Update hidden previous_price_list_rate
    item_price = frappe.get_doc("Item Price", name)
    frappe.db.set(item_price, 'previous_price_list_rate', item_price.price_list_rate)
  
    # Basic filters
    filters = {
        'item_code': item_code,
        'rm_cost_as_per': rm_cost_as_per
    }

    # Add price_list filter if provided
    if price_list:
        filters['buying_price_list'] = price_list

    # Fetch BOMs containing the item and matching criteria
    boms = frappe.get_all('BOM', filters=filters, fields=['name'])

    # Update cost for each BOM
    for bom in boms:
        bom_doc = frappe.get_doc("BOM", bom.name)
        BOM.update_cost(bom_doc, rate)
    

File: energielenker/energielenker/item_price/item_price.js
// Copyright (c) 2023, libracore AG and contributors
// For license information, please see license.txt

frappe.ui.form.on("Item Price", {
	after_save: function (frm) {
		if (cur_frm.doc.price_list_rate)  {
			if(frm.doc.price_list_rate != frm.doc.previous_price_list_rate) {
				update_bom_item_cost(frm);
			} 
		}
	},

});

function update_bom_item_cost(frm) {
    frappe.call({
        'method': "energielenker.energielenker.item_price.item_price.update_bom_item_cost",
        'args': {
			'name': frm.doc.name,
			'item_code': frm.doc.item_code,
            'rate': frm.doc.price_list_rate,
            'rm_cost_as_per': "Price List",
            'price_list': frm.doc.price_list,
            
        },
    });
}
